### ISSUE TITLE
update-locatedb.service cannot create /var/cache

### ISSUE THREAD

**Issue by colemickens** [NixOS/nixpkgs#14132]
I noticed my locatedb hasn't been updated in ~34 days.

The `update-locatedb.service` is failing. Here's the full output: https://gist.github.com/9e59c6591c07f5c824c5.

And for some reason, some of the log lines are missing from that, so here's the systemctl status output for that service: https://gist.github.com/anonymous/133677bd9f946cb62e19.

Relevant lines:

```
Mar 22 02:15:42 chimera update-locatedb-start[30204]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/home/cole/.subversion/auth': Permission denied
Mar 22 02:15:42 chimera update-locatedb-start[30204]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/home/cole/.ssh': Permission denied
Mar 22 02:15:43 chimera update-locatedb-start[30204]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/home/cole/.nixops': Permission denied
Mar 22 02:15:43 chimera update-locatedb-start[30204]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/home/cole/.docker': Permission denied
Mar 22 02:15:44 chimera update-locatedb-start[30204]: /nix/store/95wmyhk0y60r09b98l6v4vswv5m7xm4q-coreutils-8.25/bin/sort: write failed: 'standard output': Broken pipe
Mar 22 02:15:44 chimera update-locatedb-start[30204]: /nix/store/95wmyhk0y60r09b98l6v4vswv5m7xm4q-coreutils-8.25/bin/sort: write error
Mar 22 02:15:44 chimera update-locatedb-start[30204]: Failed to generate /var/cache/locatedb.n
```

**Comment by abbradar** [comment]
I can't find any trace of `update-localedb` (with a `grep`), at least on master and the latest unstable channel. What service do you mean?

**Comment by abbradar** [comment]
Argh, sorry, I read it `localedb` (notice `l`). Sorry, sleepy.

**Comment by romildo** [comment]
`update-locatedb.service` has problems on my unstable NixOS system too:
```
# systemctl -l status update-locatedb.service
● update-locatedb.service - Update Locate Database
   Loaded: loaded (/nix/store/j4syy9mzrigxxg05j15b734yp3kz1qp0-unit-update-locatedb.service/update-locatedb.service; bad; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2016-03-23 15:48:57 BRT; 13s ago
  Process: 1766 ExecStart=/nix/store/r0sji1kl4m7ypx0njw9wn5fa1klpnk43-unit-script/bin/update-locatedb-start (code=exited, status=1/FAILURE)
 Main PID: 1766 (code=exited, status=1/FAILURE)

Mar 23 15:48:56 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/alt/config/hexchat': Permission denied
Mar 23 15:48:56 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/alt/config/liferea.2014-08-06': Permission denied
Mar 23 15:48:56 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/alt/config/liferea': Permission denied
Mar 23 15:48:56 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/zcxyd6kp9hky1hici2bmdlx0zhnwf45p-findutils-4.4.2/bin/find: `/alt/lost+found': Permission denied
Mar 23 15:48:57 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/95wmyhk0y60r09b98l6v4vswv5m7xm4q-coreutils-8.25/bin/sort: write failed: 'standard output': Broken pipe
Mar 23 15:48:57 jrm.no-ip.org update-locatedb-start[1766]: /nix/store/95wmyhk0y60r09b98l6v4vswv5m7xm4q-coreutils-8.25/bin/sort: write error
Mar 23 15:48:57 jrm.no-ip.org update-locatedb-start[1766]: Failed to generate /var/cache/locatedb.n
Mar 23 15:48:57 jrm.no-ip.org systemd[1]: update-locatedb.service: Main process exited, code=exited, status=1/FAILURE
Mar 23 15:48:57 jrm.no-ip.org systemd[1]: update-locatedb.service: Unit entered failed state.
Mar 23 15:48:57 jrm.no-ip.org systemd[1]: update-locatedb.service: Failed with result 'exit-code'.
```

**Comment by DamienCassou** [comment]
I confirm the problem

**Comment by wizeman** [comment]
I have the same problem and see this in systemd's journal:
```
update-locatedb-start[5731]: /nix/store/(...)-findutils-4.4.2/bin/updatedb: line 278: /var/cache/locatedb.n: Read-only file system
```

The systemd service seems to have been defined like this:
```
ReadWriteDirectories=/var/cache/locatedb
ReadOnlyDirectories=/
```

So it seems that `ReadWriteDirectories` is not being correctly specified.

**Comment by 0xABAB** [comment]
This is what I wanted to write: 

`This was the only reason I upgraded to 16.03. Thank you.`

However, this is still not fixed: 

`locate a`

returns (after a call to `updatedb` (which has exit status 0))
```
locate: `/var/cache/locatedb': No such file or directory
```

So, this is *not* fixed in 16.03. Please reopen. If `updatedb` is not meant to be callable as root, then `updatedb` should be a wrapper of some kind which would say 

```
updatedb is not a supported command in NixOS.
```
However, I think pretty much anyone with some Linux experience wants to be able to easily build a new index when they want and `updatedb` has historically been the UI/API to do that. Changing that, if that was the intention, is a rather ill-conceived idea.

**Comment by madjestic** [comment]
So, is this supposed to be working?  I am still getting the same error.

**Comment by wizeman** [comment]
This issue was about the `update-locatedb.service` systemd service failing with a `Broken pipe` / `write error`, it wasn't about `locate` not being able to find the `locatedb` file, even after running `updatedb` manually (I would file a separate issue for that).

@madjestic is the error you are still getting the `Broken pipe` failure in the systemd service? If so, which release / branch / channel are you running?

**Comment by Chattered** [comment]
I still get the same error "read-only filesystem."

I suspect this is because the only service I have running that wants to create /var/cache is the one for updatedb, which still doesn't have write permissions to /var to do so. Manually creating /var/cache as root fixes the problem.

**Comment by peter-holm** [comment]
Just installed nixos. I to  get this errror. Is there any fix for this.?

**Comment by tischwa** [comment]
Same here on a fresh install:
# nix-channel --list
nixos https://nixos.org/channels/nixos-16.09
# updatedb 
/nix/store/w1gzi0342kiha08pym7qj09w09ffzhj4-findutils-4.6.0/bin/updatedb: line 323: /nix/store/w1gzi0342kiha08pym7qj09w09ffzhj4-findutils-4.6.0/var/locatedb.n: Read-only file system

**Comment by mwasilew2** [comment]
just installed 16.09 and I'm getting the same, Read-only file system on updatedb

**Comment by mwasilew2** [comment]
VERSION="16.09.1836.067e66a (Flounder)"
nix-env (Nix) 1.11.7


works fine if updatedb is used with --output and later locate is used with --database

when invoked by root in a login shell, updatedb tries to write to:
/nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/var/locatedb.n: Read-only file system

when invoked by user in a login shell, updatedb tries to write to:
/nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/var/locatedb.n: Read-only file system

this path comes from updatedb shell script which was modified in nix (this line is different in the gnu sources):
186: ${LOCATE_DB=/nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/var/locatedb}


systemd service seems to be updating the db:
```
# locate exim.conf
/etc/exim.conf
/nix/var/log/nix/drvs/0g/7hasymja4a0k6wvc1hnpi2mlpadil1-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/1j/m8jsca38jfa899pp489prfih91pqg0-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/ba/j0f69giq69dfyjw05b5amq9cwn56pw-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/dh/nb6l5bnkakg0blbv92lmlschh29y2r-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/km/anvw1lrp11macxvsg0wwnpspkhqzrh-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/wa/qsillqlq2d24sn4xzy2jqiv1734nsl-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/zx/iw787q71hs8nnw0yqd9m4ykb47l31v-etc-exim.conf.drv.bz2
```

but doesn't have access to some dirs:
```
# locate --database=locatedb.file exim.conf
/etc/exim.conf
/home/michal/exim.conf
/home/michal/exim.conf.no_com
/nix/store/0g7hasymja4a0k6wvc1hnpi2mlpadil1-etc-exim.conf.drv
/nix/store/1jm8jsca38jfa899pp489prfih91pqg0-etc-exim.conf.drv
/nix/store/1xxr1c5w41x0yb59vjials9wcbwv6h1p-etc/etc/exim.conf
/nix/store/6p9wws79cx2as5rqqalnvvxdk0sc6h8y-etc-exim.conf
/nix/store/75rywf64wb0qdgwhs3jwd6vny38iazd5-etc/etc/exim.conf
/nix/store/826is9r0rh9wjxlm39pjif6qsbjrdv0c-etc-exim.conf
/nix/store/baj0f69giq69dfyjw05b5amq9cwn56pw-etc-exim.conf.drv
/nix/store/c25bnxrmq8imyyvzyzi1hsivv879wckw-etc-exim.conf
/nix/store/d977s6zbj07kawddw8cmv573184zs31z-etc/etc/exim.conf
/nix/store/d9rx1dcr24ic5dnpsi7mg5iiar5ilmbw-etc-exim.conf
/nix/store/dhnb6l5bnkakg0blbv92lmlschh29y2r-etc-exim.conf.drv
/nix/store/gy08apx603iy8a1wczycmxvbn5yi8ir8-etc/etc/exim.conf
/nix/store/kmanvw1lrp11macxvsg0wwnpspkhqzrh-etc-exim.conf.drv
/nix/store/l2s6chqc06cwxrpn16kvpfw4i0rix2bb-etc-exim.conf
/nix/store/lzg0zxa923hxc5rva8nrhhwl610nkzl6-etc/etc/exim.conf
/nix/store/npn2x695vbld3awm04qkgm4z9pfm9ffv-etc/etc/exim.conf
/nix/store/pljfwlpx2chvcw5840xr7bdd3dx4h2wk-etc-exim.conf
/nix/store/waqsillqlq2d24sn4xzy2jqiv1734nsl-etc-exim.conf.drv
/nix/store/wikn3djlgsk1qscnn0zzj0q4jvc7hv35-etc/etc/exim.conf
/nix/store/xdkf7xyn3yhkpdrg86pz4sja993dljkw-etc-exim.conf
/nix/store/y5rzkhms3h2jsbbmkicjgip1qa0m2maq-fail2ban-0.9.4/etc/fail2ban/filter.d/exim.conf
/nix/store/z2w9dwybhwizlzcml6g0612b15g3nw8q-etc/etc/exim.conf
/nix/store/zxiw787q71hs8nnw0yqd9m4ykb47l31v-etc-exim.conf.drv
/nix/var/log/nix/drvs/0g/7hasymja4a0k6wvc1hnpi2mlpadil1-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/1j/m8jsca38jfa899pp489prfih91pqg0-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/ba/j0f69giq69dfyjw05b5amq9cwn56pw-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/dh/nb6l5bnkakg0blbv92lmlschh29y2r-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/km/anvw1lrp11macxvsg0wwnpspkhqzrh-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/wa/qsillqlq2d24sn4xzy2jqiv1734nsl-etc-exim.conf.drv.bz2
/nix/var/log/nix/drvs/zx/iw787q71hs8nnw0yqd9m4ykb47l31v-etc-exim.conf.drv.bz2
```

I'm not surprised it doesn't have access to home, but shouldn't the service be able to list /nix/store files?

There are no error messages for /nix/store:
```
Mar 23 19:48:43 nixos systemd[1]: Started Update Locate Database.
Mar 23 19:48:43 nixos su[4263]: Successful su for nobody by root
Mar 23 19:48:43 nixos su[4263]: pam_unix(su:session): session opened for user nobody by (uid=0)
Mar 23 19:48:43 nixos su[4265]: Successful su for nobody by root
Mar 23 19:48:43 nixos su[4265]: pam_unix(su:session): session opened for user nobody by (uid=0)
Mar 23 19:48:43 nixos su[4267]: Successful su for nobody by root
Mar 23 19:48:43 nixos su[4267]: pam_unix(su:session): session opened for user nobody by (uid=0)
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/var/spool/exim/input’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/var/spool/exim/msglog’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/var/spool/exim/db’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/var/db/sudo’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/var/lib/NetworkManager’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/lost+found’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/root’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/boot’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/sudo’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/user/1000’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/systemd/ask-password-block’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/systemd/inaccessible’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/nix/remote-stores’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/run/keys’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/etc/NetworkManager/system-connections’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/home/guest’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/home/michal’: Permission denied
Mar 23 19:48:43 nixos update-locatedb-start[4229]: /nix/store/95vfigaqdg8jg5bk961k1x06j86p5fh0-findutils-4.6.0/bin/find: ‘/home/lost+found’: Permission denied
```


the ticket is about service failing to write to /var/cache/locatedb, but it works fine on 16.09, so I think it can be closed.

I see two other problems though:
- update-locatedb service not being able to process /nix/store
- running updatedb on the cli tries to write to /nix/store which is a read-only file system

**Comment by calbrecht** [comment]
#24249 is caused by an option, now named services.locate.prunePaths, which contains "/nix/store" by default, the issue is closed already.
#24250 is resolved by merged #23152, the issue can be closed.
#17397 is about using updatedb from cli and missing /var/cache directory and needs a decision.
